src/utils.ts

```diff
@@ -173,15 +173,27 @@ export function areMapsEqual(
   let isValueEqual = a.size === b.size;
 
   if (isValueEqual && a.size) {
+    const matchedIndices: Record<number, true> = {};
```

Comment:
I like the solution! 
Another way would be to transform Maps and Sets to arrays to compare them. It would mean less code to maintain and possible performance gains (going from On² to On complexity) but a big change in behavior, especially for the data passed to the custom comparator... I mention all that, but you most likely already thought about it while implementing the Map and Set comparators in the first place, so feel free to ignore it.

Reply:
I actually used to have it this way, and in my perf testing this was actually more performant. Granted, that was a year ago; I definitely look to re-examine this over time, but not within the scope of this PR.